arangod/Aql/CountingMemoryResource.h

```diff
@@ -0,0 +1,59 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2014-2023 ArangoDB GmbH, Cologne, Germany
+/// Copyright 2004-2014 triAGENS GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Aditya Mukhopadhyay
+////////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#include <memory_resource>
+
+#include "Basics/ResourceUsage.h"
+
+namespace arangodb::aql {
+struct CountingMemoryResource : std::pmr::memory_resource {
+  CountingMemoryResource(memory_resource* base,
+                         ResourceMonitor& resourceMonitor)
+      : base(base), _resourceMonitor(resourceMonitor) {}
+
+ private:
+  void* do_allocate(size_t bytes, size_t alignment) override {
+    void* mem = base->allocate(bytes, alignment);
+    _resourceMonitor.increaseMemoryUsage(bytes);
+
+    return mem;
```

Comment:
What if `increaseMemoryUsage(...)` throws here?
Wouldn't that leak the memory acquired by the `allocate(...)` call?

Reply:
As far as I can tell, I could do one of two things to address this:

1. Swap the order of statements so as to update the resource monitor first and then perform the `allocate`. However, `allocate` could also throw at this point and I'm not really sure what kind of cleanup is required if it does.
2. I could wrap the resource monitor update in a try-catch block, and deallocate in the catch block (followed by a rethrow).

Not sure which, if either, is a good fix.

Reply:
I think we need to wrap the second statement in a try catch, so if it goes wrong we can roll back the first one.
Only then it will be safe.